### PR TITLE
KEYCLOAK-16616 authenticationSessions map in RootAuthenticationSessio…

### PR DESCRIPTION
**Pull request by martin-kanis**
…nEntity grows unboundedly

### CODE REVIEW

**Comment by martin-kanis** [comment]
@mhajas @andymunro This piece of documentation is needed for Keycloak 14. If you could review it that would be great. Thanks!

**Review by mhajas — COMMENTED** [review]
Thank you for the PR @martin-kanis. I am missing some introduction to the issue. Should we briefly add something about what types of sessions Keycloak have and when they are created and what it is causing when there is too many authenticationSessions? And then continue with mitigations for RootAuthenticationSession and then authenticationSession? Or we assume that the reader knows this already?

**Comment by martin-kanis** [comment]
> Thank you for the PR @martin-kanis. I am missing some introduction to the issue. Should we briefly add something about what types of sessions Keycloak have and when they are created and what it is causing when there is too many authenticationSessions? And then continue with mitigations for RootAuthenticationSession and then authenticationSession? Or we assume that the reader knows this already?

This kind of introduction would probably belong to a different section than "Threat Mitigation" (Not sure where though). However I slightly expanded the description of auth sessions, threats and I also mentioned that it affects only deployments that use infinispan provider for auth sessions.

**Review by mhajas — DISMISSED** [review]
Thanks, @martin-kanis, this is exactly what I imagined.

**Review by mhajas — DISMISSED** [review]

File: server_admin/topics/threat/auth-sessions-limit.adoc [hunk 1]
```diff
@@ -0,0 +1,37 @@
+=== Limit Authentication Sessions
+
+{project_name} creates an authentication session for each authentication request within a realm. This session is internally represented by `RootAuthenticationSession`. Each `RootAuthenticationSession` can have multiple authentication sessions for a given client. In a browser authentication flow the browser session translates to `RootAuthenticationSession` while browser's tabs translate to a collection of authentication sessions within the `RootAuthenticationSession`.
```
Comment (andymunro) [inline server_admin/topics/threat/auth-sessions-limit.adoc]: Add a comma before "the browser session"
Remove the 's after browser

File: server_admin/topics/threat/auth-sessions-limit.adoc [hunk 2]
```diff
@@ -0,0 +1,37 @@
+=== Limit Authentication Sessions
+
+{project_name} creates an authentication session for each authentication request within a realm. This session is internally represented by `RootAuthenticationSession`. Each `RootAuthenticationSession` can have multiple authentication sessions for a given client. In a browser authentication flow the browser session translates to `RootAuthenticationSession` while browser's tabs translate to a collection of authentication sessions within the `RootAuthenticationSession`.  
+
+NOTE: This section describes deployments that use the infinispan provider for authentication sessions.
+ 
+Higher memory usage and a network overload in a cluster may occur for deployments where there's many active `RootAuthenticationSession` with a lot of authentication sessions.
```
Comment (andymunro) [inline server_admin/topics/threat/auth-sessions-limit.adoc]: change "there's" to "there are"

File: server_admin/topics/threat/auth-sessions-limit.adoc [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  
+There's a possibility to mitigate the aforementioned threats. The maximum allowed number of `RootAuthenticationSession` should be addressed by limiting the request rate per time on the proxy or load balancer in front of {project_name}. Maximum number of authentication sessions per `RootAuthenticationSession` can be configured in `authenticationSessions` SPI by setting property `authSessionsLimit`. The default value is set to 300 authentication sessions per a `RootAuthenticationSession`. When this limit is reached the oldest authentication session will be removed after a new authentication session request.
```
Comment (andymunro) [inline server_admin/topics/threat/auth-sessions-limit.adoc]: Change "There's" to "There is"
Add "The" before "Maximum"
Add a comma after "When this limit is reached"

**Review by andymunro — CHANGES_REQUESTED** [review]
A few changes

File: server_admin/topics/threat/auth-sessions-limit.adoc [hunk 4]
```diff
@@ -0,0 +1,37 @@
+=== Limit Authentication Sessions
+
+{project_name} creates an authentication session for each authentication request within a realm. This session is internally represented by `RootAuthenticationSession`. Each `RootAuthenticationSession` can have multiple authentication sessions for a given client. In a browser authentication flow the browser session translates to `RootAuthenticationSession` while browser's tabs translate to a collection of authentication sessions within the `RootAuthenticationSession`.  
+
+NOTE: This section describes deployments that use the infinispan provider for authentication sessions.
```
Comment (hmlnarik) [inline server_admin/topics/threat/auth-sessions-limit.adoc]:
```suggestion
NOTE: This section describes deployments that use the `{jdgserver_name}` provider for authentication sessions.
```

**Review by hmlnarik — APPROVED** [review]

**Pull request merged**